src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CrlCache.cs

```diff
@@ -0,0 +1,253 @@
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal static class CrlCache
+    {
+        public static void AddCrlForCertificate(
+            X509Certificate2 cert,
+            SafeX509StoreHandle store,
+            X509RevocationMode revocationMode,
+            DateTime verificationTime,
+            ref TimeSpan remainingDownloadTime)
+        {
+            // In Offline mode, accept any cached CRL we have.
+            // "CRL is Expired" is a better match for Offline than "Could not find CRL"
+            if (revocationMode != X509RevocationMode.Online)
+            {
+                verificationTime = DateTime.MinValue;
+            }
+
+            if (AddCachedCrl(cert, store, verificationTime))
+            {
+                return;
+            }
+
+            // Don't do any work if we're over limit or prohibited from fetching new CRLs
+            if (remainingDownloadTime <= TimeSpan.Zero ||
+                revocationMode != X509RevocationMode.Online)
+            {
+                return;
+            }
+
+            DownloadAndAddCrl(cert, store, ref remainingDownloadTime);
+        }
+
+        private static bool AddCachedCrl(X509Certificate2 cert, SafeX509StoreHandle store, DateTime verificationTime)
+        {
+            string crlFile = GetCachedCrlPath(cert);
+
+            using (SafeBioHandle bio = Interop.libcrypto.BIO_new_file(crlFile, "rb"))
+            {
+                if (bio.IsInvalid)
+                {
+                    return false;
+                }
+
+                // X509_STORE_add_crl will increase the refcount on the CRL object, so we should still
+                // dispose our copy.
+                using (SafeX509CrlHandle crl = Interop.libcrypto.PEM_read_bio_X509_CRL(bio))
+                {
+                    if (crl.IsInvalid)
+                    {
+                        return false;
+                    }
+
+                    // If crl.LastUpdate is in the past, downloading a new version isn't really going
+                    // to help, since we can't rewind the Internet. So this is just going to fail, but
+                    // at least it can fail without using the network.
+                    //
+                    // If crl.NextUpdate is in the past, try downloading a newer version.
+                    DateTime nextUpdate = OpenSslX509CertificateReader.ExtractValidityDateTime(
+                        Interop.Crypto.GetX509CrlNextUpdate(crl));
+
+                    if (nextUpdate < verificationTime)
+                    {
+                        return false;
+                    }
+
+                    if (!Interop.libcrypto.X509_STORE_add_crl(store, crl))
+                    {
+                        //error:0B07D065:x509 certificate routines:X509_STORE_add_crl:cert already in hash table
+                        //throw Interop.libcrypto.CreateOpenSslCryptographicException();
```

Comment:
There are three options that I see:

* Remove the if, ignore the function result (effectively what I currently have)
* Try to determine the current OpenSSL error code without creating an exception
    * The way that OpenSSL's error queue works that feels slightly problematic
* Attach the error code to the exception object, and here do something like `CryptographicException e = CreateException(); if (e.HResult != X509_R_CERT_ALREADY_IN_HASH_TABLE) { throw e; }`

The first has a certain distaste to it, the second might be non-destructively-achievable with some native interop, and the third puts values into the HResult field that weren't HResult values.

And I'm sort of torn on which one I like the most...